src/dex/nerve/nerve-math.ts

```diff
@@ -0,0 +1,317 @@
+import { biginterify, ZERO, ONE, MathUtil } from './utils';
+import { Logger } from '../../types';
+import { PoolState, ReadonlyOrWritable } from './types';
+
+export class NervePoolMath {
+  readonly FEE_DENOMINATOR = biginterify(10 ** 10);
+  readonly A_PRECISION = biginterify(100);
+  readonly MAX_LOOP_LIMIT = biginterify(256);
+  readonly POOL_PRECISION_DECIMALS = biginterify(18);
+
+  constructor(protected name: string, protected logger: Logger) {}
+
+  calculateSwap(
+    state: PoolState,
+    tokenIndexFrom: number,
+    tokenIndexTo: number,
+    dx: bigint,
+    blockTimestamp: bigint,
+  ) {
+    const xp = this._xp(state);
+
+    // uint256 x = dx.mul(self.tokenPrecisionMultipliers[tokenIndexFrom])
+    //    .add(xp[tokenIndexFrom]);
+    const x =
+      dx * state.tokenPrecisionMultipliers[tokenIndexFrom] + xp[tokenIndexFrom];
+
+    const y = this._getY(
+      state,
+      tokenIndexFrom,
+      tokenIndexTo,
+      x,
+      xp,
+      blockTimestamp,
+    );
+
+    if (y === null) {
+      state.isValid = false;
+      return null;
+    }
+
+    // dy = xp[tokenIndexTo].sub(y).sub(1);
+    let dy = xp[tokenIndexTo] - y - ONE;
+
+    // dyFee = dy.mul(self.swapFee).div(FEE_DENOMINATOR);
+    const dyFee = (dy * state.swapFee) / this.FEE_DENOMINATOR;
+
+    // dy = dy.sub(dyFee).div(self.tokenPrecisionMultipliers[tokenIndexTo]);
+    dy = (dy - dyFee) / state.tokenPrecisionMultipliers[tokenIndexTo];
+    return { dy, dyFee };
+  }
+
+  calculateWithdrawOneToken(
+    state: PoolState,
+    tokenAmount: bigint,
+    tokenIndex: number,
+    blockTimeStamp: bigint,
+  ) {
+    const out = this._calculateWithdrawOneTokenDY(
+      state,
+      tokenIndex,
+      tokenAmount,
+      blockTimeStamp,
+    );
+    if (out === null) return null;
+    let { dy, newY } = out;
+
+    // uint256 dySwapFee = _xp(self)[tokenIndex].sub(newY)
+    //  .div(self.tokenPrecisionMultipliers[tokenIndex]).sub(dy)
+    const dySwapFee =
+      (this._xp(state)[tokenIndex] - newY) /
+        state.tokenPrecisionMultipliers[tokenIndex] -
+      dy;
+
+    // dy = dy.mul(FEE_DENOMINATOR.sub(calculateCurrentWithdrawFee(self, account)))
+    //  .div(FEE_DENOMINATOR);
+    dy =
+      (dy * (this.FEE_DENOMINATOR - this._calculateCurrentWithdrawFee(state))) /
+      this.FEE_DENOMINATOR;
+
+    return { dy, dyFee: dySwapFee };
+  }
+
+  swapUnderlying(state: PoolState) {}
```

Comment:
is it suppose to be here?

Reply:
Its a small step in the future when we support meta pools :)
Removed